### PR TITLE
feat(debugger): add dynamic span function probes

### PR DESCRIPTION
**Pull request by shatzi**
Adding support for Dynamic-Span function probes that decorate any function with a span, called dynamic span.

Dynamic span have the following setting:
1. span name is "dd.dynamic.span"
2. resource name is the qualified function name
3. all probes tags are added to the span
4. a special "debugger.probeid" tag is also added with the probe id.
5. exception thrown from the function are attached to the span error

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Author is aware of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

File: ddtrace/debugging/_capture/dynamic_span.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        if not isinstance(self.probe, SpanFunctionProbe):
+            return
+
+        probe = self.probe
+
+        _args = dict(self.args) if self.args else {}
+        if not self._eval_condition(_args):
+            return
+
+        self.span = ddtrace.tracer.trace(SPAN_NAME, resource=probe.func_qname)
```
[the remark this answers is not in the document]
Comment (shatzi) [reply ddtrace/debugging/_capture/dynamic_span.py:37]: I think this is the same a JP have it dd-trace-java. maybe we need to add package as well...

File: tests/debugging/test_debugger.py [hunk 2]
```diff
@@ -947,3 +947,85 @@ def test_debugger_log_live_probe_generate_messages():
...
+                )
+            )
+
+            mutator(arg=[42])  # should not trigger span
+
+            self.assert_span_count(0)
+
+            mutator(arg=[])  # should trigger span
+
+            self.assert_span_count(1)
```
Comment (Kyle-Verhoog) [inline tests/debugging/test_debugger.py:1004]: 👍 nice, I like these tests, they're very readable!

File: tests/debugging/test_debugger.py [hunk 3]
```diff
@@ -947,3 +947,85 @@ def test_debugger_log_live_probe_generate_messages():
...
+            (span,) = self.get_spans()
+
+            assert span.name == "dd.dynamic.span"
+            assert span.resource == "mutator"
+            assert span.get_tags()["debugger.probeid"] == "span-probe"
+
+    def test_debugger_snap_probe_linked_to_parent_span(self):
```
Comment (Kyle-Verhoog) [inline tests/debugging/test_debugger.py:1011]: can we extend this test case to also have mutator create a span to ensure the dynamic span becomes the parent of spans created within the instrumented function? It's pretty crucial that we assert this behaviour as well.

**Review by Kyle-Verhoog — COMMENTED** [review]
Looks great, good stuff @shatzi 👏 

Once my comments are addressed then it is good to go from me.

File: ddtrace/debugging/_capture/dynamic_span.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        if not self._eval_condition(dict(self.args) if self.args else {}):
+            return
+
+        self._span_cm = ddtrace.tracer.trace(
+            SPAN_NAME,
+            service=None,  # Currently unused
+            resource=probe.func_qname,
+            span_type=None,  # Currently unused
+        )
+        span = self._span_cm.__enter__()
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/debugging/_capture/dynamic_span.py:46]: > is there a reason we call __enter__?

Yes! This is to make it future-proof, in light of changes like #5122, by not assuming that `Tracer.trace()` returns a span object, but rather a `ContextManager[Span]`.

File: ddtrace/debugging/_probe/remoteconfig.py [hunk 5]
```diff
@@ -108,12 +109,16 @@ def _wrapper(*args, **kwargs):
 
 
 def _create_probe_based_on_location(args, attribs, line_class, function_class):
-    # type: (Dict[str, Any], Dict[str, Any], Type, Type) -> Any
+    # type: (Dict[str, Any], Dict[str, Any], Optional[Type], Optional[Type]) -> Any
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/debugging/_probe/remoteconfig.py:112]: We don't have a class for span line probes so we would pass `None` in that case (see below). Within this function we now check against `None`.

File: tests/debugging/test_debugger.py (hunk 3 above)
Comment (P403n1x87) [reply tests/debugging/test_debugger.py:1012]: I have added a new test case below with a dynamic span as root. I hope this is what you asked! 🙂

**Pull request merged**